folding/src/checker.rs

```diff
@@ -0,0 +1,283 @@
+//! A kind of pseudo-prover, will compute the expressions over the witness a check row by row
+//! for a zero result.
+
+use crate::{
+    expressions::{FoldingColumnTrait, FoldingCompatibleExpr, FoldingCompatibleExprInner},
+    instance_witness::Instance,
+    ExpExtension, FoldingConfig, Radix2EvaluationDomain, RelaxedInstance, RelaxedWitness, Sponge,
+};
+use ark_ec::AffineCurve;
+use ark_ff::{Field, Zero};
+use ark_poly::Evaluations;
+use kimchi::{
+    circuits::{expr::Variable, gate::CurrOrNext},
+    curve::KimchiCurve,
+};
+use mina_poseidon::{
+    constants::PlonkSpongeConstantsKimchi,
+    sponge::{DefaultFqSponge, ScalarChallenge},
+    FqSponge,
+};
+use poly_commitment::PolyComm;
+use std::ops::Index;
+
+#[cfg(not(test))]
+use log::debug;
+#[cfg(test)]
+use std::println as debug;
+
+// 0. We start by defining the field and the curve that will be used in the
+// constraint system, in addition to the sponge that will be used to generate
+// challenges.
+pub type Fp = ark_bn254::Fr;
+pub type Curve = ark_bn254::G1Affine;
+pub type SpongeParams = PlonkSpongeConstantsKimchi;
+pub type BaseSponge = DefaultFqSponge<ark_bn254::g1::Parameters, SpongeParams>;
+
+// 1. We continue by defining a generic type of columns and selectors.
+// The selectors can be seen as additional (public) columns that are not part of
+// the witness.
+// The column must implement the trait [Hash] as it will be used by internal
+// structures of the library.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum Column {
+    X(usize),
+    Selector(usize),
+}
+
+// 2. We implement the trait [FoldingColumnTrait] that allows to distinguish
+// between the public and private inputs, often called the "instances" and the
+// "witnesses".
+// By default, we consider that the columns are all witness values and selectors
+// are public.
+impl FoldingColumnTrait for Column {
+    fn is_witness(&self) -> bool {
+        match self {
+            Column::X(_) => true,
+            Column::Selector(_) => false,
+        }
+    }
+}
+
+// TODO: get rid of trait Sponge in folding, and use the one from kimchi
+impl Sponge<Curve> for BaseSponge {
```

Comment:
Note that it will be removed in https://github.com/o1-labs/proof-systems/pull/2180/